### PR TITLE
Fix #9878 - ChillerHeater:Absorption:DoubleEffect should calculate Specific heat of Water and Air at the right temperatures

### LINKED ISSUE #9878: ChillerHeater:Absorption:DoubleEffect heating model uses 0 C for calculating specific heat of water
Issue overview
--------------
In calcHeater the specific heat of water is calculated before the return water temperature is initialized such that Cp_HW is always based on a hot water temperature of 0 C. Found during #9877 style changes.

ChillerExhaustAbsorption.cc

    void ExhaustAbsorberSpecs::calcHeater()

        Real lHotWaterReturnTemp(0.0);

        Cp_HW = FluidProperties::GetSpecificHeatGlycol(
        state, state.dataPlnt->PlantLoop(LoopNum).FluidName, lHotWaterReturnTemp, state.dataPlnt->PlantLoop(LoopNum).FluidIndex, RoutineName);

        lHotWaterReturnTemp = state.dataLoopNodes->Node(lHeatReturnNodeNum).Temp;

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (ExhFiredAbsorptionChiller.idf)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #9878

I have been careful to make my commits in the right order and with the right separation so it's easy to follow what's going on.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/ChillerExhaustAbsorption.unit.cc [hunk 1]
```diff
@@ -637,29 +638,60 @@ TEST_F(EnergyPlusFixture, ExhAbsorption_calcHeater_Fix_Test)
...
+    state->dataLoopNodes->Node(thisChillerHeater.HeatSupplyNodeNum).TempSetPoint = hwSupplySetpoint;
+
+    // Minimum temperature leaving the Chiller absorber is 176.6 C (350 F)
+    constexpr Real64 exhaustInTemp = 350.0;
+    constexpr Real64 absLeavingTemp = 176.667;
+    constexpr Real64 exhaustInMassFlowRate = 0.5;
+    constexpr Real64 exhaustInHumRate = 0.005;
+    state->dataLoopNodes->Node(thisChillerHeater.ExhaustAirInletNodeNum).Temp = exhaustInTemp;
+    state->dataLoopNodes->Node(thisChillerHeater.ExhaustAirInletNodeNum).MassFlowRate = exhaustInMassFlowRate;
+    state->dataLoopNodes->Node(thisChillerHeater.ExhaustAirInletNodeNum).HumRat = exhaustInHumRate;
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ChillerExhaustAbsorption.unit.cc:682]: Setting the HumRat to non zero, so we catch the fact that CpAir was always calculated at 0.

File: tst/EnergyPlus/unit/ChillerExhaustAbsorption.unit.cc [hunk 2]
```diff
@@ -637,29 +638,60 @@ TEST_F(EnergyPlusFixture, ExhAbsorption_calcHeater_Fix_Test)
...
+    Real64 loadinput = 5000.0;
+    bool const runflaginput = true;
     thisChillerHeater.calcHeater(*state, loadinput, runflaginput);
 
-    EXPECT_NEAR(thisChillerHeater.HeatingLoad, 21085.0, 1e-6);
+    const Real64 CpHW = FluidProperties::GetSpecificHeatGlycol(*state, hwPlantLoop.FluidName, hwReturnTemp, hwPlantLoop.FluidIndex, "UnitTest");
+    EXPECT_EQ(4185.0, CpHW);
+    const Real64 expectedHeatingLoad = (hwSupplySetpoint - hwReturnTemp) * hwMassFlow * CpHW;
+
+    EXPECT_NEAR(20925.0, expectedHeatingLoad, 1e-6);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ChillerExhaustAbsorption.unit.cc:692]: Recalculate the expected heating load properly, with a CpHW that is properly calculated at 60°C and not at 0°C always.

File: tst/EnergyPlus/unit/ChillerExhaustAbsorption.unit.cc [hunk 3]
```diff
@@ -669,7 +701,11 @@ TEST_F(EnergyPlusFixture, ExhAbsorption_calcHeater_Fix_Test)
     EXPECT_NEAR(thisChillerHeater.ElectricPower, 400.0, 1e-6);
     EXPECT_NEAR(thisChillerHeater.ExhaustInTemp, 350.0, 1e-6);
     EXPECT_NEAR(thisChillerHeater.ExhaustInFlow, 0.5, 1e-6);
-    EXPECT_NEAR(thisChillerHeater.ExhHeatRecPotentialHeat, 87087.5769469, 1e-6);
+
+    Real64 const CpAir = Psychrometrics::PsyCpAirFnW(exhaustInHumRate);
+    Real64 const epectedExhHeatRecPotentialHeat = exhaustInMassFlowRate * CpAir * (exhaustInTemp - absLeavingTemp);
+    EXPECT_NEAR(87891.51, epectedExhHeatRecPotentialHeat, 0.01);
+    EXPECT_NEAR(epectedExhHeatRecPotentialHeat, thisChillerHeater.ExhHeatRecPotentialHeat, 0.01);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ChillerExhaustAbsorption.unit.cc]: Check the expectedExhHeatRecPotentialHeat

```suggestion
    Real64 const CpAir = Psychrometrics::PsyCpAirFnW(exhaustInHumRate);
    Real64 const expectedExhHeatRecPotentialHeat = exhaustInMassFlowRate * CpAir * (exhaustInTemp - absLeavingTemp);
    EXPECT_NEAR(87891.51, expectedExhHeatRecPotentialHeat, 0.01);
    EXPECT_NEAR(expectedExhHeatRecPotentialHeat, thisChillerHeater.ExhHeatRecPotentialHeat, 0.01);
```

File: src/EnergyPlus/ChillerExhaustAbsorption.cc [hunk 4]
```diff
@@ -1909,6 +1909,7 @@ void ExhaustAbsorberSpecs::calcHeater(EnergyPlusData &state, Real64 &MyLoad, boo
     int LoopNum = this->HWPlantLoc.loopNum;
     DataPlant::LoopSideLocation LoopSideNum = this->HWPlantLoc.loopSideNum;
 
+    lHotWaterReturnTemp = state.dataLoopNodes->Node(lHeatReturnNodeNum).Temp;
```
Comment (jmarrec) [inline src/EnergyPlus/ChillerExhaustAbsorption.cc]: Minimal fix for Cp_HW: make sure HW return Temp is initialized before the GetSpecificHeatGlycol is called

File: src/EnergyPlus/ChillerExhaustAbsorption.cc [hunk 5]
```diff
@@ -2007,6 +2007,7 @@ void ExhaustAbsorberSpecs::calcHeater(EnergyPlusData &state, Real64 &MyLoad, boo
 
         lExhaustInTemp = state.dataLoopNodes->Node(lExhaustAirInletNodeNum).Temp;
         lExhaustInFlow = state.dataLoopNodes->Node(lExhaustAirInletNodeNum).MassFlowRate;
+        lExhaustAirHumRat = state.dataLoopNodes->Node(this->ExhaustAirInletNodeNum).HumRat;
```
Comment (jmarrec) [inline src/EnergyPlus/ChillerExhaustAbsorption.cc]: Minimal Fix for CpAir: initialize lExhaustAirHumRat

File: src/EnergyPlus/ChillerExhaustAbsorption.cc [hunk 6]
```diff
@@ -1874,86 +1874,64 @@ void ExhaustAbsorberSpecs::calcHeater(EnergyPlusData &state, Real64 &MyLoad, boo
...
+            assert(false);
+            return 0.0;
+        }
+        }
+    }();
+
+    auto const &heatReturnNode = state.dataLoopNodes->Node(this->HeatReturnNodeNum);
+    Real64 const HeatDeltaTemp = std::abs(heatReturnNode.Temp - HeatSupplySetPointTemp);
+    // reporting: hot water mass flow rate
+    Real64 lHotWaterMassFlowRate = heatReturnNode.MassFlowRate;
```
Comment (jmarrec) [inline src/EnergyPlus/ChillerExhaustAbsorption.cc:1908]: Refactor the routine to limit the number of declared variables atop the file.

**Comment by jmarrec** [comment]
ExhFiredAbsorptionChiller.idf has, quite logically, some diffs. This is expected, and this is quite small.

Summing for entire year, then comparing:

image

Absolute:




image


Relative (%):

image

File: src/EnergyPlus/ChillerExhaustAbsorption.cc (hunk 4 above)
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerExhaustAbsorption.cc]: Yep that's a good move :+1:

File: src/EnergyPlus/ChillerExhaustAbsorption.cc (hunk 5 above)
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerExhaustAbsorption.cc]: Yep :+1:

**Review by Myoldmopar — APPROVED** [review]
Minimal fixes look good, and thanks for the refactor and test.  Checking it locally now.

**Comment by Myoldmopar** [comment]
Decent CI looks as expected -- I would certainly think diffs would pop up in the applicable absorption chiller model.  Otherwise clean.  Built and tested with develop locally, all good, merging, thanks @jmarrec

**Pull request merged**